### ISSUE TITLE
Active, background codecs shouldn't be reclaimed.

### ISSUE THREAD

**Issue by jamespearce2006** [w3c/webcodecs#363]
While I have no problem with the user agent reclaiming inactive codecs, I have some significant misgivings about active video decoders being reclaimed from background tabs.

The idea of reclaiming video decoders from inactive tabs seems to follow the current behaviour for \<video>. It makes sense to reclaim decoders used by \<video> because, by their nature, these render to the screen - if the page isn't visible, why keep decoding? However, that's not necessarily true in the WebCodec case. One of the nice things about this API is that it unlocks a variety of new use cases, many of which we haven't thought of, that no longer rely on the traditional fetch-decode-render pipeline. It seems a shame, therefore, to enforce the same resource-management rules that don't necessarily fit these new workflows.

The argument is that the user can still hear the audio output from an inactive tab, but can't see the video output. This assumes 
that the only purpose of decoding video is to render a picture to the page immediately. I can think of a number of uses of a video decoder that don't do this. For example, extracting burned-in timecode, or detecting faces. It's likely a user would switch tabs while waiting for the video to be processed and then wonder why their processing has failed or stalled. A more contrived example might be one in which the video is being used to generate sound or avant-garde music. In that case, the audio would stop when the user switches tabs.

I would also argue that a bad actor can overcome the decoder reclamation anyway, by creating a video encoder and continually feeding it with trivial, inconsequential data.

Is there any reason why a background tab can't just have a shorter idle timeout instead?

**Comment by sandersdan** [comment]
>  I have some significant misgivings about active video decoders being reclaimed from background tabs.

Me too, the problem is that we don't have a better proposal.

There are some tasks, such as background transcoding, that are currently out of scope for WebCodecs because they would need other web platform work to be viable (eg. a background compute type of web worker).

The `<video>` and WebRTC like cases are the most straightforward to fully support now. The reality is that we don't currently have a proposal for supporting additional use cases that is convenient for developers, aligns incentives with users, and supports every important use case.

> I would also argue that a bad actor can overcome the decoder reclamation anyway, by creating a video encoder and continually feeding it with trivial, inconsequential data.

This is true. I advocate for giving more control to the UA to close codecs, but we also need a consistent developer experience across browsers and we have not yet found a better way to be both predictable and useful.

> Is there any reason why a background tab can't just have a shorter idle timeout instead?

This question concerns specific implementations of WebCodecs, so I'll offer the reasoning we have used in the past for `<video>` on Chrome Android. In this case there is likely to be a limited number of platform codecs available, and allowing pages to hold codecs in the background can prevent other apps the user is actively using from functioning correctly. Therefore Chrome Android aggressively closes background decoding sessions. This interferes with temporarily switching away from a tab being used for WebRTC conferencing, so there is a heuristic exception when encoders are active.

Chrome Android can be more lenient if the codec is a software codec, since that is less likely to interfere with other apps. (Note though that Android will still suspend background apps eventually, a separate API is still necessary to hold compute resources in the background on Android.)

**Comment by jamespearce2006** [comment]
Thanks @sandersdan, I understand it's a bit of a balancing act.
I have some questions - sorry if these are naïve.

I know you can only talk about implementation details for Chrome, but does reclamation occur if there is no pressure on decoders? For example, if I'm on a desktop PC, will Chrome reclaim active decoders on background tabs even if it doesn't immediately need to do so? I'm just trying to gauge the likely frequency this will happen for our users.

One of the main concerns I have about reclamation of active decoders is the fact they can be reclaimed mid-GOP, which makes it much harder to recover and retry. Is there anything that can be done to try and reduce the likelihood of this happening? Perhaps by holding off the reclamation until after the next flush if the previous flush was less than n milliseconds ago (ie, if the application is regularly flushing). The application could then try and reacquire a decoder for the next GOP...

You mentioned that browsers can be more lenient with software decoders. Would it be possible for browsers to dynamically change a VideoDecoder instance from being hardware-backed to software-backed if the hardware decoder needs to be reclaimed? Ultimately, the Javascript VideoDecoder is an abstract representation of a decoder, so I don't believe there's any requirement that the underlying resources are fixed.

Finally, WebRTC use cases tend to get more leniency because they've had to request permission from the user via GUM. What about a user prompt to get permission for background decoding? Something like "This page wants to decode video in the background"?

**Comment by sandersdan** [comment]
> but does reclamation occur if there is no pressure on decoders?

Yes, we implemented the behavior consistently across platforms in Chrome, primarily so that the developer experience would be consistent.

> One of the main concerns I have about reclamation of active decoders is the fact they can be reclaimed mid-GOP

It's generally best to feed only a few frames at a time into the decoder, so as long as there is active playback this won't be a concern. It does become a concern for paused video; the only recovery option will be to restart decoding at the previous keyframe.

Chrome's `<video>` element also does this, but has an exception for content with long average keyframe intervals and some streaming sources on Desktop. WebCodecs is not currently eligible for those exceptions.

> Would it be possible for browsers to dynamically change a VideoDecoder instance from being hardware-backed to software-backed if the hardware decoder needs to be reclaimed?

This is possible at keyframe boundaries if the `VideoDecoderConfig` doesn't explicitly prefer hardware acceleration. We can't guarantee that a keyframe will arrive at the right time though so it's not a solution in general.

> What about a user prompt to get permission for background decoding?

I don't think a prompt is reasonable, since it's a technical concern that most users won't have an opinion on. Notably though on Android the only way to guarantee background compute resources is to have a persistent notification, which serves a similar purpose. This is what I would expect would happen for the "background compute" worker I mentioned earlier.
